### PR TITLE
Api integration

### PR DESCRIPTION
**Pull request by Tushar3099**
Integration of APIs with the revamped code structure.
These APIs are working for now : 

- [x] `/api/v1/reload`
- [x] `/get-machines`
- [x] `/get-domain-entities
- [x] implement step

### CODE REVIEW

**Comment by Tushar3099** [comment]
@Harkishen-Singh Please review the PR. Currently, this PR enables the APIs given in the description.

File: src/lib/api/api.go [hunk 1]
```diff
@@ -24,12 +37,120 @@ func (a *API) Router() *mux.Router {
...
+	a.mutex.Lock()
+	a.domainMap = d
+	log.Info("msg", "Updated API domain map")
+	a.mutex.Unlock()
+}
+
 func (a *API) RegisterRoutes() {
 	a.router.HandleFunc("/api/v1/reload", a.Reload)
+	a.router.HandleFunc("/get-machines", a.getMachines)
+	a.router.HandleFunc("/get-domain-entities", a.getDomainEntity)
```
Comment (ruddi10) [inline src/lib/api/api.go:54]: the route need to be prefixed with /api/v1/

Comment (Tushar3099) [reply src/lib/api/api.go:54]: Okay, will do the changes!

File: tsdb/querier/querier.go [hunk 2]
```diff
@@ -208,11 +207,11 @@ func (q *Query) exec(blockStream []tsdb.Block, encoding bool) (i interface{}) {
 		return q.ReturnNILResponse()
 	}
 	// decode the selected range of blocks
-	blocksDecoder := decode.NewBlockDecoding(resultingBlockSlice[0].Type)
+	// blocksDecoder := decode.NewBlockDecoding(resultingBlockSlice[0].Type)
```
Comment (Harkishen-Singh) [inline tsdb/querier/querier.go:210]: Why did we comment these out?

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]
Thank you for the implementation. Some work needs to be done, but overall looks good.

**Comment by Tushar3099** [comment]
@Harkishen-Singh, I have completed the Querier package with unit tests and also integrated the querier with Decoder. The integration of the Querier with API and other proposed changes will be done in subsequent commits. I will try to do it ASAP.

**Comment by Tushar3099** [comment]
@Harkishen-Singh, I have completed the API integration with the querier package handling almost all the edge cases. Please give a review of the recent changes. I will commit the earlier proposed changes with the new changes after review.

File: src/lib/api/api.go [hunk 3]
```diff
@@ -24,12 +47,212 @@ func (a *API) Router() *mux.Router {
...
+				a.send(w, makeError(err), http.StatusBadRequest)
+				return
+			}
+			qryRes, err := qry.Exec()
+			if err != nil {
+				a.send(w, makeError(err), http.StatusBadRequest)
+				return
+			}
+			var status string
+			if qryRes.Values[0].Value.(evaluate.Response).Status != 200 {
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:124]:
```suggestion
			if qryRes.Values[0].Value.(evaluate.Response).Status != 200 { // We verify if status is 200, and only then ask the UI to show `UP` for the route.
```

File: src/lib/api/api.go [hunk 4]
```diff
@@ -24,12 +47,212 @@ func (a *API) Router() *mux.Router {
...
+}
+
+func (a *API) getDomainEntity(w http.ResponseWriter, r *http.Request) {
+	domain_or_ip := r.URL.Query().Get("domain_or_ip")
+	type data struct {
+		ChainName  string `json:"chain_name"`
+		EntityName string `json:"entity_name"`
+		Status     string `json:"status"`
+	}
+	res := []data{}
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:92]:
```suggestion
	res := make([]data, 0)
```

File: src/lib/api/api.go [hunk 5]
```diff
@@ -24,12 +47,212 @@ func (a *API) Router() *mux.Router {
...
+	}
+
+	qry, err := querier.New(querier.TypeRange, path, start, end)
+	if err != nil {
+		a.send(w, makeError(err), http.StatusBadRequest)
+		return
+	}
+	qryRes, err := qry.Exec()
+	if err != nil {
+		a.send(w, makeError(err), http.StatusBadRequest)
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:172]:
```suggestion
		a.send(w, makeError(err), http.InternalServerError)
```

File: src/lib/api/api.go [hunk 6]
```diff
@@ -24,12 +47,212 @@ func (a *API) Router() *mux.Router {
...
+		return
+	}
+	if ok := tsdb.VerifyChainPathExists(path); !ok {
+		a.send(w, makeError(fmt.Errorf("%s", "INVALID_PATH")), http.StatusBadRequest)
+		return
+	}
+
+	qry, err := querier.New(querier.TypeRange, path, start, end)
+	if err != nil {
+		a.send(w, makeError(err), http.StatusBadRequest)
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:167]: code 500

File: src/lib/api/api.go [hunk 7]
```diff
@@ -24,12 +47,212 @@ func (a *API) Router() *mux.Router {
...
+		Unit string  `json:"unit"`
+		Data []block `json:"data"`
+	}
+	var res data
+	for i := 0; i < len(qryRes.Values); i += (1 + step) {
+		switch qryRes.Type {
+		case "jitter":
+			res.Unit = "ms"
+			res.Data = append(res.Data, block{
+				Timestamp: qryRes.Values[i].NormalizedTime / int64(time.Millisecond),
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:191]: please recheck this

File: tsdb/file/chain.go [hunk 8]
```diff
@@ -136,6 +136,14 @@ func loadBlocks(raw []byte) []Block {
 	return stream
 }
 
+func FetchChainStream(path string) ([]Block, error) {
```
Comment (Harkishen-Singh) [inline tsdb/file/chain.go:139]: optional: make the []Block as slice

File: tsdb/file/querier/querier.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+	// given query range and vice-versa for lastBlockIndex.
+	// Also firstBlockTime represents the oldest block time in the stream
+	// and lastBlockTime represents the latest block time in the stream.
+	var firstBlockIndex, lastBlockIndex int
+	firstBlockTime, lastBlockTime := (*stream)[0].NormalizedTime, (*stream)[streamLen-1].NormalizedTime
+	switch q.typ {
+	case TypeFirst:
+		return []tsdb.Block{(*stream)[streamLen-1]}, nil
+	case TypeRange:
+
```
Comment (Harkishen-Singh) [inline tsdb/file/querier/querier.go:124]:
```suggestion
```

File: tsdb/file/querier/querier.go [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+func (q *Query) Exec() (*QueryResponse, error) {
+	bstream, err := tsdb.FetchChainStream(q.dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("loading chain error: %s", err)
+	}
+	q.timestamp = time.Now()
+	sortedStream := sortStream(bstream)
+	resStream, err := q.fetchBlocks(sortedStream)
```
Comment (Harkishen-Singh) [inline tsdb/file/querier/querier.go:73]: please add a comment on what fetchBlock does.

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]

File: tsdb/querier/querier.go (hunk 2 above)
Comment (Tushar3099) [reply tsdb/querier/querier.go:210]: They were creating errors that's why I commented on them.

**Review by Harkishen-Singh — APPROVED** [review]

**Pull request merged**